src/lib/transaction_database/transaction_database.ml

```diff
@@ -1,113 +1,390 @@
+        match Time.compare time1 time2 with
+        | 0 ->
+            Transaction.compare txn1 txn2
+        | x ->
+            x
     end
 
-    module Latest = V1
+    include T
+    include Comparable.Make (T)
   end
 
-  type t = Stable.Latest.t =
-    {sender: Public_key.Compressed.t option; receiver: Public_key.Compressed.t}
+  type cache =
+    { user_transactions: Txn_with_date.Set.t Public_key.Compressed.Table.t
+    ; all_transactions: Time.t Transaction.Table.t }
+
+  type t = {database: Database.t; cache: cache}
+
+  let create ?directory_name () =
+    let directory =
+      match directory_name with
+      | None ->
+          Uuid.to_string (Uuid_unix.create ())
+      | Some name ->
+          name
+    in
+    { database= Database.create ~directory
+    ; cache=
+        { user_transactions= Public_key.Compressed.Table.create ()
+        ; all_transactions= Transaction.Table.create () } }
+
+  (* TODO: make load function #2333 *)
+
+  let close {database; _} = Database.close database
+
+  let get_participants (transaction : Transaction.t) =
+    match transaction with
+    | Fee_transfer (One (pk, _)) ->
+        [pk]
+    | Fee_transfer (Two ((pk1, _), (pk2, _))) ->
+        [pk1; pk2]
+    | Coinbase {Coinbase.proposer; fee_transfer; _} ->
+        Option.value_map fee_transfer ~default:[proposer] ~f:(fun (pk, _) ->
+            [proposer; pk] )
+    | User_command checked_user_command ->
+        let user_command = User_command.forget_check checked_user_command in
+        let sender = User_command.sender user_command in
+        let payload = User_command.payload user_command in
+        let receiver =
+          match User_command_payload.body payload with
+          | Stake_delegation (Set_delegate {new_delegate}) ->
+              new_delegate
+          | Payment {receiver; _} ->
+              receiver
+        in
+        [receiver; sender]
+
+  let add {database; cache= {all_transactions; user_transactions}} transaction
+      date =
+    if Option.is_none (Hashtbl.find all_transactions transaction) then (
+      Database.set database ~key:transaction ~data:date ;
+      Hashtbl.add_exn all_transactions ~key:transaction ~data:date ;
+      List.iter (get_participants transaction) ~f:(fun pk ->
+          let user_txns =
+            Option.value
+              (Hashtbl.find user_transactions pk)
+              ~default:Txn_with_date.Set.empty
+          in
+          let user_txns' =
+            Txn_with_date.Set.add user_txns (transaction, date)
+          in
+          Hashtbl.set user_transactions ~key:pk ~data:user_txns' ) )
+
+  let get_transactions {cache= {user_transactions; _}; _} public_key =
+    let open Option in
+    value ~default:[]
+    @@ ( Hashtbl.find user_transactions public_key
+       >>| Txn_with_date.Set.to_list
+       >>| List.map ~f:(fun (txn, _) -> txn) )
+
+  let get_pagination_query ~f t public_key transaction =
+    (let open Option.Let_syntax in
```

Comment:
can you make a local let binding here, it's a bit confusing to apply the `|>` to the whole parenthetical expression